Make buffer::reinterpret more standard conforming.

fodinabor opened the pull request:
- return buffer with rebound allocator,
- add overload without range parameter,
- allow and handle differently sized target types,
- throw an exception if the byte size is not fully covered.

Handling differently sized types is a bit tricky as hipSYCL's page mechanism works on the original element's size. Thus, whenever the size of the element type of a ranged accessor is different than the size of the type, the buffer was originally created with, all pages of the buffer are migrated to prevent partial values.

fodinabor committed:
Make buffer::reinterpret more standard conforming.

- return buffer with rebound allocator,
- add overload without range parameter,
- allow and handle different sized target types,
- throw an exception if the byte size is not fully covered.

fodinabor commented:
@hipsycl-ci-bot test this change again

fodinabor committed:
Mention impact of reinterpreting a buffer in runtime-spec.

illuhad commented on tests/sycl/buffer.cpp:
```diff
@@ -65,28 +65,79 @@ BOOST_AUTO_TEST_CASE(buffer_versioning) {
+  s::buffer<std::uint16_t, 1, s::buffer_allocator<std::uint16_t>> buf_d = buf_a.reinterpret<std::uint16_t>(s::range<1>{32});
+  s::buffer<std::uint32_t, 1, s::buffer_allocator<std::uint32_t>> buf_e = buf_a.reinterpret<std::uint32_t, 1>();
+  s::buffer<std::uint16_t, 2, s::buffer_allocator<std::uint16_t>> buf_f = buf_a.reinterpret<std::uint16_t>(s::range<2>{4, 8});
+  s::buffer<std::uint32_t, 2, s::buffer_allocator<std::uint32_t>> buf_g = buf_a.reinterpret<std::uint32_t, 2>();
```
Some line breaks?

illuhad commented on include/hipSYCL/sycl/libkernel/accessor.hpp:
```diff
@@ -1097,19 +1098,21 @@ class accessor
-      auto explicit_requirement =
-          rt::make_operation<rt::buffer_memory_requirement>(
+      std::unique_ptr<rt::buffer_memory_requirement> explicit_requirement;
+      if(sizeof(dataT) != data->get_num_elements())
```
```suggestion
      if(sizeof(dataT) != data->get_element_size())
```
I don't understand this - shouldn't it be as in the suggestion?

illuhad commented on include/hipSYCL/sycl/libkernel/accessor.hpp:
```diff
@@ -1097,19 +1098,21 @@ class accessor
+      std::unique_ptr<rt::buffer_memory_requirement> explicit_requirement;
+      if(sizeof(dataT) != data->get_num_elements())
+        explicit_requirement.reset(new rt::buffer_memory_requirement(
+              data, rt::id<3>{}, data->get_num_elements(),
```
Since only offset and range are different, would it make sense to only find the right offset/range inside the `if` and construct the requirement outside, to share some more code between the code paths?

illuhad commented on include/hipSYCL/sycl/handler.hpp:
```diff
@@ -105,20 +106,22 @@ class handler {
+      req.reset(new rt::buffer_memory_requirement(
         data.mem, rt::make_id(data.offset), rt::make_range(data.range), mode,
-        AccessorType::access_target);
+        AccessorType::access_target));
```
We have this logic now in multiple places. I wonder if it might be a good idea to have a single `get_effective_offset/range` or similar somewhere, so that we only need to change a single place if we want to optimize this case some more?

illuhad commented on doc/runtime-spec.md:
```diff
@@ -49,10 +49,12 @@ For each allocation managed on each device, the `buffer` implementation shall tr
 
 Data transfers generated from accessors (see below) shall cause transferred pages to be marked as up-to-date on the target allocation.
 
+After reinterpreting a `buffer` with a differing type, the implementation may require the transfer of all pages, despite a smaller range being specified by an accessor.
```
```suggestion
If a `buffer` is reinterpreted to a data type of different size than the original buffer element size, the implementation may assume a page range for accessors to the reinterpreted buffer that is larger than the page range as defined above.
```

fodinabor committed:
Apply suggestions from illuhad's code review.

Co-authored-by: Aksel Alpay <aksel.alpay@uni-heidelberg.de>

fodinabor committed:
Unify getting effective offset/range for buffer req.

Also fixes wrong checks and follow up issues.
After reshaping the buffer, it is also rather difficult to calculate the required pages correctly, thus require all pages for now.
